### ISSUE TITLE
Retaining exceptions for mail content on close

### ISSUE THREAD

**Issue by demided** [HorlogeSkynet/thunderbird-user.js#23]
Hi,
Since I use TB for professional purposes, I have a professional signature that includes a logo that is pulled from my website. In TB's settings, I therefore go to Privacy & Security > Mail content (with the "allow..." checkbox unchecked) > Exception, and I add the relevant exception.
The problem is that this exception gets flushed on close. Since I cannot seem to find anything pref directly relating to this, I get this is a consequence of sanitize on close.
Is there a way to keep those exceptions despite sanitizing on close? Most of what I find in the user.js seems to relate to cookies, etc., things for browsing from TB (web content), but not really mail content.
Any idea of what could help here?
Thanks!
Dem.

**Comment by HorlogeSkynet** [comment]
Hey @demided,

From what I see in the source code, such parameter is indeed not saved as a "preference" (thus, not stored in `prefs.js`).
If this has something to do with auto-sanitization on close, you could try what these comments tell : https://github.com/HorlogeSkynet/thunderbird-user.js/blob/master/user.js#L846

So, I guess, this would bring us to something like :

```js
user_pref("network.cookie.lifetimePolicy", 2);

user_pref("privacy.clearOnShutdown.cookies", false);
user_pref("privacy.clearOnShutdown.offlineApps", false);
```

Could you try it out and tell me/us if this worked for you ? 🙂 

Thanks, bye 👋

**Comment by demided** [comment]
Hi @HorlogeSkynet, thanks for the quick reply and for the suggestion. Unfortunately, no, it has not worked. I was actually skeptical about the cookie part, since the user.js mentions it as relating to web_content and not mail_content. Either way, the suggestion does not prevent the deletion of exceptions.
PS: and, to be clear, I did make sure to close TB, make the changes, re-open, re-add the URLs, re-close, and re-re-open to check ;)

**Comment by HorlogeSkynet** [comment]
Hey back @demided,

Have you try to disable sanitation on close _completely_ ?

```js
user_pref("privacy.sanitize.sanitizeOnShutdown", false);
```

If this works, there is somehow an item that include these exceptions during deletion.
I've found yesterday where they are handled in TB source code, but I couldn't yet figure _where_ they get written on disk.

Bye :wave:

**Comment by demided** [comment]
Ok, so I tried disabling sanitation on close entirely and it did not work BUT it led me to realise that, for some reason that eludes me, the "network.cookie.lifetimePolicy" and "privacy.sanitize.sanitizeOnShutdown" prefs do not stick from the user.js. The first one is always at 0 and the second always on true. 
This is confusing because these prefs are placed, as far as I know, in the right section of the user.js -- almost all the way at the bottom of the section copied from the user-overrides, just where the custom code should be (see image below).
I checked and the file is saved properly and it's the one in the right TB profile. Not only that, but when I change "premissions.default.image" (which you see just above in the picture) from 3 to 2 and back to 3, these changes actually happen in TB as they should (and all my other custom prefs work as well). 
Is there something that could override these prefs?

<img width="735" alt="Schermata 2022-07-12 alle 23 14 13" src="https://user-images.githubusercontent.com/106995386/178597099-3b547562-2440-439f-b9d3-a9bcf33f1f3d.png">

**Comment by demided** [comment]
I had not tried, it's true, but the all structure seemed sound and, indeed, the value is "Success! User settings loaded". To be sure, any custom setting above/apart from the ones just added following your recommendation are taken into account successfully. It's just these ones that don't stick. 
We agree that, like for Arkenfox, all the prefs are handled in order and that anything down the file will overwrite the same pref(s) placed high up, right?

**Comment by demided** [comment]
PS: checking the prefs.js file confirms this: the value changed in the user.js (privacy.sanitize.sanitizeOnShutdown and privacy.clearOnShutdown.cookies/offlineApps marked as false) are still set to true.

**Comment by demided** [comment]
PPS: almost unsurprisingly, even changing the privacy.sanitize.sanitizeOnShutdown to false in the main part of the user.js (or commenting all instances) still keeps it at true after TB relaunch, so something else is making it true.

**Comment by HorlogeSkynet** [comment]
Please check these references, some behaviors have changed with TB 101/102 :

1. https://searchfox.org/mozilla-central/source/netwerk/cookie/CookieService.cpp#56
2. https://github.com/arkenfox/user.js/commit/3fcc711c5acc24032fb9ee5d6b104f6dde0a2315
3. https://bugzilla.mozilla.org/buglist.cgi?bug_id=1681493,1681495,1681498,1759665

You'll want to completely remove `network.cookie.lifetimePolicy` in order to prevent (1) migration to be performed, re-enabling sanitation on close in this case.

In the future, could you please fill the GitHub issue template in ? It'd spare me some time when looking at the code.

**TL; DR** : As stated in #20, this script ~~might not be~~ isn't ready for Thunderbird 102, yet.

Bye :wave:

**Comment by HorlogeSkynet** [comment]
Please see #24, still a work in progress.

**Comment by demided** [comment]
Ok, sorry about the template thing, might have helped, yes. This being said, removing network.cookie.lifetimepolicy did allow the prefs to be changed, but did not solve the issue of the mail content exceptions to be retained.

**Comment by HorlogeSkynet** [comment]
Good first step ! What about disabling `privacy.sanitize.sanitizeOnShutdown` now ? 🙄

**Comment by demided** [comment]
Like I said, now when turning it to false, it is indeed marked as false. But that doesn't solve the issue. Same with all the various "clearOnShutdown".

**Comment by HorlogeSkynet** [comment]
(I'm sorry Dem but I cannot follow precisely what you tried or not remotely 🙁)

Anyway, the other day I managed to track down where/how these exceptions are stored in TB source code, until [this](https://searchfox.org/comm-esr102/source/mail/components/preferences/permissions.js#22).

FYI, I managed to reproduce this issue with TB v91, so I'll revert this issue renaming.

While looking at preferences that could deal with SQLite and InnoDB, I've come across :
```js
/* 5004: disable permissions manager from writing to disk [FF41+] [RESTART]
 * [NOTE] This means any permission changes are session only
 * [1] https://bugzilla.mozilla.org/967812 ***/
user_pref("permissions.memory_only", true); // [HIDDEN PREF]
```
... which actually disables reading **and writing** of permissions on disk.

Switching it back fixed this issue in my case (without dealing to auto-sanitation). Could you try it out on your side ?

Thanks, bye 👋

**Comment by demided** [comment]
Very interesting, and amazing work! This did it. So, indeed, no link to sanitize or clearOnShutdown (I did have a hunch that this was for web content and not mail content); but turning this to false actually brought back a whole lot of elements (URLs and email addresses) that I had previously allowed and that were not long picked up. I removed all of them but now the work URL is saved and stayed after close. Great great job, thanks a bunch!
